tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc

```diff
@@ -559,6 +560,295 @@ LogicalResult GetElementShapeFromResultType(
+                           SmallVectorImpl<int64_t>& shape) {
+  shape.clear();
+  auto eltShape = element_shape.getDefiningOp<TF::ConstOp>();
+  auto listSize = list_size.getDefiningOp<TF::ConstOp>();
+  if (!eltShape || !listSize) return;
+
+  ElementsAttr eltShapeAttr = eltShape.value();
+  ElementsAttr listSizeAttr = listSize.value();
+
+  // Shape of the tensor is listSize followed by element_shape.
+  shape.push_back(
+      listSizeAttr.getValue(llvm::None).cast<IntegerAttr>().getInt());
+  for (unsigned i = 0, m = eltShapeAttr.getNumElements(); i < m; i++) {
+    shape.push_back(eltShapeAttr.getValue({i}).cast<IntegerAttr>().getInt());
+  }
+}
+
+/// Find unique while op users of the result of EmptyTensorListOp.
+static void findUserWhileOps(Value result, DenseSet<Operation*>& whileOpUsers) {
+  for (Operation* useOp : result.getUsers()) {
+    if (isa<TF::WhileOp>(useOp)) whileOpUsers.insert(useOp);
+  }
+}
+
+/// Returns true if EmptyTensorListOp within WhileOp's body has exactly one
+/// PushBack use.
+static bool checkExactlyOnePushbackUse(TF::EmptyTensorListOp op,
+                                       TF::WhileOp useOp, FuncOp body_branch) {
+  Block& body_block = body_branch.front();
+  Value emptyTensor = op.getResult();
+  SmallVector<unsigned, 2> indicesUsed;
+  for (unsigned i = 0, n = useOp.getNumOperands(); i < n; i++) {
+    if (useOp.getOperand(i) == emptyTensor) indicesUsed.push_back(i);
+  }
+
+  // Traverse through all the indices in WhileOp's operand where
+  // EmptyTensorListOp is used and check the uses of EmptyTensorListOp.
+  for (unsigned index : indicesUsed) {
+    auto argumentUsers = body_block.getArgument(index).getUsers();
+    unsigned countPushBackOps = 0;
+    for (Operation* user : argumentUsers) {
+      if (isa<TF::TensorListPushBackOp>(user)) countPushBackOps++;
+      // We bail out on rewrite if there are nested WhileOps.
+      else if (isa<TF::WhileOp>(user))
+        return false;
+    }
+    if (countPushBackOps != 1) return false;
+  }
+
+  return true;
+}
+
+/// Utility function that returns true if argUsers contains a Sub/Add Op with
+/// 1 as an operand.
+static bool isLoopIVWithStrideOne(Value::user_range argUsers) {
+  for (Operation* use : argUsers) {
+    if (isa<TF::AddV2Op, TF::SubOp>(use)) {
+      Value stride = use->getOperand(1);
+      auto constOp = stride.getDefiningOp<TF::ConstOp>();
+      ElementsAttr constAttr = constOp.value();
+      // We handle cases where the loopIV is changed by +-1.
+      if (constAttr.getValue(llvm::None).cast<IntegerAttr>().getInt() == 1)
+        return true;
+    }
+  }
+  return false;
+}
+
+/// Returns the maximum number of tensor elements that'll be pushed in the
+/// EmptyTensorList. In case more than one push_back operation is applied on
+/// the EmptyTensorList within the same WhileOp body, we return None and thus
+/// the rewriter fails in those cases.
+static Optional<unsigned> findDynamicShape(TF::EmptyTensorListOp op,
```

Comment:
What happens if there are ops like TensorListPushBackBatch, TensorListConcatV2 and others? There might be some new ops in the future or some custom tensor list ops by the users.

Reply:
That's a very good point @smit-hinsu .
Currently the code only deals with `TensorListPushBack` ops with an added constraint that only 1 push back per iteration is allowed : `checkExactlyOnePushbackUse`.

The constraint can be relaxed gradually in subsequent PRs and I've added a TODO for handling `TensorListPushBackBatch`, `TensorListConcatV2`, etc. One way to deal with them in the subsequent PRs is to perhaps add an OpTrait in such ops that changed the shape of the list and go about with it.

Reply:
That will make this transformation incorrect. Trait can handle existing operations but someone could be using custom operations that we don't know about.